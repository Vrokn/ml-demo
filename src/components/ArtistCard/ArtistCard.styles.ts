import styled from "@emotion/styled";

interface CardFrontProps {
  backgroundImage: string;
}

export const CardDescription = styled.div(() => ({
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'center',
  gap: '1em',
  minHeight: '40vh',
  padding: '0 2em',
  width: "70vw",
  "@media (min-width: 768px)": {
    width: "70vw",
  },
  "@media (min-width: 900px)": {
    width: "28vw",
  },
  "@media (min-width: 1200px)": {
    width: "20vw",
  },
}));

export const CardFront = styled.div<CardFrontProps>(props => ({
  minHeight: '40vh',
  overflow: 'hidden',
  display: 'flex',
  flexGrow: '1',
  backgroundImage: `url(${props.backgroundImage})`,
  alignItems: 'end',
  justifyContent: 'center',
  backgroundPosition: 'center',
  backgroundSize: 'cover',
  borderRadius: '4px',
  padding: '0 2em',
  width: "70vw",
  "@media (min-width: 768px)": {
    width: "70vw",
  },
  "@media (min-width: 900px)": {
    width: "28vw",
  },
  "@media (min-width: 1200px)": {
    width: "20vw",
  },
}));
