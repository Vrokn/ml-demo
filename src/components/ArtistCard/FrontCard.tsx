import React from "react";
import { Typography } from "@mui/material";
import { CardFront } from "./ArtistCard.styles";

const FrontCard = ({ name, image }: { name: string; image: string; }) => (
  <CardFront backgroundImage={image}>
    <Typography
      variant="h3"
      style={{
        textShadow: '0px 0px 20px black',
        WebkitTextStroke: '2px rgb(0,0,0,0.1)',
        color: 'white',
        fontWeight: "bold",
        padding: '0.7em 0',
        textAlign: 'center'
      }}
    >
      {name.toLocaleUpperCase()}
    </Typography>
  </CardFront>
);

export default FrontCard;