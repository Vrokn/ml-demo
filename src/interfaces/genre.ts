
export interface Genre {
  id: number;
  name: string;
  parent_id?: number;
  is_primary?: number;
}
