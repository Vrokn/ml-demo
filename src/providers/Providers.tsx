import { ReactNode } from "react";
import { BrowserRouter } from "react-router-dom";
import { ArtistProvider } from "../context/SavedArtistContext/SavedArtistContext";
import { SnackbarProvider } from "../context/SnackbarContext/SnackbarContext";
import { theme } from "./theme/theme";
import { ThemeProvider } from "@mui/material";

export const Providers = ({ children }: { children: ReactNode; }) => {
  return (
    <BrowserRouter>
      <SnackbarProvider>
        <ArtistProvider>
          <ThemeProvider theme={theme}>
            {children}
          </ThemeProvider>
        </ArtistProvider>
      </SnackbarProvider>
    </BrowserRouter>
  );
};
