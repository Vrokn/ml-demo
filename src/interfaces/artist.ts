import { Genre } from "./genre";

export interface Artist {
  id: number;
  name: string;
  genres: Genre[];
  popularity: number;
  image: string;
}
