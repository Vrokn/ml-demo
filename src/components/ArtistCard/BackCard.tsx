import React from "react";
import { CardDescription } from "./ArtistCard.styles";
import { Button, Link, Typography } from "@mui/material";
import GenreChips from "../GenreChips/GenreChips";
import { Artist } from "../../interfaces/artist";
import { Genre } from "../../interfaces/genre";
import StarBorderIcon from '@mui/icons-material/StarBorder';
import StarIcon from '@mui/icons-material/Star';

interface BackCardProps {
  artist: Artist;
  primaryGenres: Genre[];
  secondaryGenres: Genre[];
  isPresent: boolean;
  handleClick: () => void;
}

const BackCard = ({ artist, primaryGenres, secondaryGenres, isPresent, handleClick }: BackCardProps) => (
  <CardDescription>
    <Link href={`/artist/${artist.id}`}>
      <Typography variant="h6" color='primary' fontWeight="bold">
        {artist.name.toLocaleUpperCase()}
      </Typography>
    </Link>
    {primaryGenres.length > 0 && (
      <GenreChips
        genres={primaryGenres}
        label={"Primary"}
        color={"primary"}
      />
    )}
    {secondaryGenres.length > 0 && (
      <GenreChips
        genres={secondaryGenres}
        label={"Secondary"}
        color={"secondary"}
      />
    )}
    <Button
      children={isPresent ? "Remove" : "Add"}
      onClick={handleClick}
      startIcon={isPresent ? <StarIcon /> : <StarBorderIcon />}
      variant={isPresent ? "contained" : "outlined"}
    />
  </CardDescription>
);

export default BackCard;