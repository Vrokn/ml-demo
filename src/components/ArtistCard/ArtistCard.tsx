import React, { useState } from "react";
import { useSpring, a } from "@react-spring/web";
import { Box } from "@mui/material";
import { useArtistContext } from "../../context/SavedArtistContext/SavedArtistContext";
import { Artist } from "../../interfaces/artist";
import BackCard from "./BackCard";
import FrontCard from "./FrontCard";

interface ArtistCardProps {
  artist: Artist;
}

const ArtistCard: React.FC<ArtistCardProps> = (props: ArtistCardProps) => {
  const [flipped, setFlipped] = useState(false);
  const { transform, opacity } = useSpring({
    opacity: flipped ? 1 : 0,
    transform: `perspective(600px) rotateX(${flipped ? 180 : 0}deg)`,
    config: { mass: 5, tension: 500, friction: 80 }
  });
  const { artist } = props;
  const { artists, toggleArtist } = useArtistContext();

  const handleClick = () => {
    toggleArtist(artist);
  };

  const isPresent = artists.some(a => a.id === artist.id);

  const primaryGenres = artist.genres.filter(item => item.is_primary);
  const secondaryGenres = artist.genres.filter(item => item.is_primary !== 1);

  return (
    <Box
      onMouseEnter={() => setFlipped(true)}
      onMouseLeave={() => setFlipped(false)}
      display="flex"
    >
      <a.div
        style={{
          willChange: 'transform, opacity',
          position: "relative",
          opacity: opacity.to((o) => 1 - o),
          transform,
          flex: 1,
        }}
      >
        <FrontCard name={artist.name} image={artist.image} />
      </a.div>
      <a.div
        style={{
          willChange: 'transform, opacity',
          position: "absolute",
          opacity,
          transform,
          rotateX: "180deg",
          flex: 1,
        }}
      >
        <BackCard
          artist={artist}
          primaryGenres={primaryGenres}
          secondaryGenres={secondaryGenres}
          isPresent={isPresent}
          handleClick={handleClick}
        />
      </a.div >
    </Box >
  );
};

export default ArtistCard;
