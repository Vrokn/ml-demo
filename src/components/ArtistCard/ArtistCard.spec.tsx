import React from 'react';
import { render } from '@testing-library/react';
import ArtistCard from './ArtistCard';
import { Artist } from '../../interfaces/artist';
import { Providers } from '../../providers/Providers';

const mockArtist: Artist = {
  id: 1,
  name: 'Artist',
  genres: [{ id: 1, name: 'Test Genre', parent_id: 1 }],
  popularity: 50,
  image: 'https://test.image.url',
};

describe('ArtistCard', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Providers>
        <ArtistCard artist={mockArtist} />
      </Providers>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
