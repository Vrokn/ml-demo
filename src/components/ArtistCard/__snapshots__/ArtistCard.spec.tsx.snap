// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-k008qs"
  >
    <div
      style="will-change: transform, opacity; position: relative; opacity: 1; transform: perspective(600px) rotateX(0deg); flex: 1;"
    >
      <div
        class="css-1kumyd4"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 css-z8x0v9-MuiTypography-root"
          style="text-shadow: 0px 0px 20px black; color: white; font-weight: bold; padding: 0.7em 0px; text-align: center;"
        >
          ARTIST
        </h3>
      </div>
    </div>
    <div
      style="will-change: transform, opacity; position: absolute; opacity: 0; transform: perspective(600px) rotateX(0deg) rotateX(180deg); flex: 1;"
    >
      <div
        class="css-jlnxgi"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-7mmmvh-MuiTypography-root-MuiLink-root"
          href="/artist/1"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-twns0k-MuiTypography-root"
          >
            ARTIST
          </h6>
        </a>
        <div
          class="css-4z29z6"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1vo1ilw-MuiTypography-root"
          >
            Secondary:
          </span>
          <div
            class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorSecondary MuiChip-filledSecondary css-1xjc7rt-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
            >
              Test Genre
            </span>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-13yqcsl-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m22 9.24-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
          Add
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
